File: java/src/main/java/benchmarks/Matmul1T.java
package benchmarks;

public class Matmul1T extends Benchmark {
    private int n;
    private long resultVal;

    public Matmul1T() {
        n = (int) configVal("n");
        resultVal = 0L;
    }

    @Override
    public String name() {
        return "Matmul1T";
    }

    private double[][] matmul(double[][] a, double[][] b) {
        int m = a.length;
        int n = a[0].length;
        int p = b[0].length;

        double[][] b2 = new double[n][p];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < p; j++) {
                b2[j][i] = b[i][j];
            }
        }

        double[][] c = new double[m][p];
        for (int i = 0; i < m; i++) {
            double[] ci = c[i];
            double[] ai = a[i];
            for (int j = 0; j < p; j++) {
                double s = 0.0;
                double[] b2j = b2[j];
                for (int k = 0; k < n; k++) {
                    s += ai[k] * b2j[k];
                }
                ci[j] = s;
            }
        }

        return c;
    }

    private double[][] matgen(int n) {
        double tmp = 1.0 / n / n;
        double[][] a = new double[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = tmp * (i - j) * (i + j);
            }
        }

        return a;
    }

    @Override
    public void run(int iterationId) {
        double[][] a = matgen(n);
        double[][] b = matgen(n);
        double[][] c = matmul(a, b);

        resultVal += Helper.checksumF64(c[n >> 1][n >> 1]);
    }

    @Override
    public long checksum() {
        return resultVal;
    }
}

File: kotlin/src/main/kotlin/benchmarks/Matmul1T.kt
package benchmarks

import Benchmark

class Matmul1T : Benchmark() {
    private var n: Long = 0
    private var resultVal: UInt = 0u

    init {
        n = configVal("n")
    }

    private fun matmul(
        a: Array<DoubleArray>,
        b: Array<DoubleArray>,
    ): Array<DoubleArray> {
        val m = a.size
        val n = a[0].size
        val p = b[0].size

        val b2 = Array(p) { DoubleArray(n) }
        for (i in 0 until n) {
            for (j in 0 until p) {
                b2[j][i] = b[i][j]
            }
        }

        val c = Array(m) { DoubleArray(p) }
        for (i in 0 until m) {
            val ai = a[i]
            val ci = c[i]
            for (j in 0 until p) {
                var s = 0.0
                val b2j = b2[j]
                for (k in 0 until n) {
                    s += ai[k] * b2j[k]
                }
                ci[j] = s
            }
        }
        return c
    }

    private fun matgen(n: Int): Array<DoubleArray> {
        val tmp = 1.0 / n / n
        val a = Array(n) { DoubleArray(n) }

        for (i in 0 until n) {
            for (j in 0 until n) {
                a[i][j] = tmp * (i - j) * (i + j)
            }
        }
        return a
    }

    override fun run(iterationId: Int) {
        val a = matgen(n.toInt())
        val b = matgen(n.toInt())
        val c = matmul(a, b)
        val center = c[(n shr 1).toInt()][(n shr 1).toInt()]
        resultVal += Helper.checksumF64(center)
    }

    override fun checksum(): UInt = resultVal

    override fun name(): String = "Matmul1T"
}
